### ISSUE TITLE
Hook for server console commands

### ISSUE THREAD

**Issue by JavidPack** [bluemagic123/tModLoader#60]
Forum Suggestion:

> Whilst its been mentioned. Can we have more options on the server controller? I used to use a server controller that gave access to starting season events or starting invasions and such.
> Any chance this could get added to Tmodloader as it helps when making a fun server/Video series

**Comment by Chicken-Bones** [comment]
I approve of this in theory, it fits with tModLoader's default mod scope

**Comment by Murfalo** [comment]
What do you think would be the best way to approach this?  I was thinking along the lines of adding a ServerInput hook to Mod

**Comment by Chicken-Bones** [comment]
Agreed

**Comment by JavidPack** [comment]
I was thinking registering a name, description, and an Action<string> in Load, and adding a number to any duplicate command so we can avoid problems with duplicate command names, but doing it just like chatinput is fine as well.

**Comment by Murfalo** [comment]
I think that's a great idea!  Is there a reason why ChatInput doesn't follow the same approach?

Expanding on your idea, here's a rough sketch/pseudocode:
```c#
class Load:
    void AddServerInput(string text, string description, Func<string, string, void> action)
    void AddChatInput(string text, ref bool broadcast, Func<string, ref bool, void> action)
```
This makes life much easier for things such as dealing with duplicates, implementing a /help command, etc.  If this looks good to you all I'll get started working on this!  Any feedback is appreciated.

**Comment by Chicken-Bones** [comment]
Func<string, string, void> is Action<string, string>, and ref requires a delegate
What are the arguments supposed to be?
How do you handle commands with parameters?
What about duplicate commands?

**Comment by Murfalo** [comment]
Hey @Chicken-Bones! I'm kind of a C# noob so basically the Func stuff was just me trying to use function pointers and my brain dying at the same time.  Basically someone making a mod would do something like:

```c#
void SpawnNPC(string text):
    <Spawn NPC based on string>

void Load():
    AddChatInput("npc", false, SpawnNPC);
```
This would make it so that when a player called the /npc command, the inputted string would be passed to the SpawnNPC function and SpawnNPC would handle it and do an action accordingly.  Something I personally think would be nice would be if tModLoader parsed the parameters and passed them as a string[] to the function (SpawnNPC in this case) which would make writing short, clean code a lot easier for the mod programmer.

Internally, when a mod calls AddChatInput/AddServerInput the information about that input would be stored within a data structure in memory which makes life much easier when it comes to writing code to make decisions about that data.

TLDR:
I'm a C# noob and don't know much about delegates but that can be fixed using the power of Google.
1) What are the arguments supposed to be? string[] args
2) tModLoader will call the command's action with an array of strings, up to the action to handle that
3) Having command info in an internal data structure makes dealing with duplicates easier.  For example,  after loading is complete can check if there are duplicates and display a warning or append a number to the command name and notify the players somehow.

**Comment by Chicken-Bones** [comment]
decent progress
what's the purpose of broadcast with a ChatCommand, give me a use case

**Comment by Murfalo** [comment]
Oh, broadcast is just to decide whether or not the command will show up in chat.  So if it's false the command is hidden so when you say something like /useless 10 10 it doesn't show up in chat where other players can see.

**Comment by Chicken-Bones** [comment]
I advise providing some utility functions/frameworks for the following cases
1. replying to a client command (maybe just Main.addChatText or whatever it's called is fine)
2. client broadcast should be something like, boolean show, which shows it just to the player that sent it
3. replying to the client who typed the command on the server (for server side management mods)
4. sending messages to all clients on the server (say if someone changes the time of day)
5. server broadcast should probably only show the command to the client (so they have history)

**Comment by JavidPack** [comment]
I think we forgot to close this.
